From 93d774a5e32c1170860cdedfbe1268a93dc0f6c6 Mon Sep 17 00:00:00 2001
From: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Date: Wed, 31 Jan 2024 14:39:23 +0530
Subject: [PATCH] wifi: ath12k: fix mac address assignment in split-phy

Currently during mac register, mac address assigned to ieee80211
hardware is ab->mac_addr. In non MLO case, this results in same mac
address assigned for two different split-phy radios that are within
the same chip. This happens because in split-phy two PHYs share the
same ath12k_base structure.

Thus, fix this issue by assigning ar->mac_addr in non-MLO case and
ab->mac_addr in MLO case to ieee80211 hardware during mac register.

Signed-off-by: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Signed-off-by: Raj Kumar Bhagat <quic_rajkbhag@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index aa9562c..0d08d54 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -15557,6 +15557,7 @@ static int ath12k_mac_hw_register(struct ath12k_hw *ah)
 	int i, j, ret;
 	u32 antennas_rx, antennas_tx;
 	bool unregister = false;
+	u8 *mac_addr = NULL;
 	static const u32 cipher_suites[] = {
 		WLAN_CIPHER_SUITE_TKIP,
 		WLAN_CIPHER_SUITE_CCMP,
@@ -15607,6 +15608,15 @@ static int ath12k_mac_hw_register(struct ath12k_hw *ah)
 
 		hw->wiphy->max_ap_assoc_sta += ar->max_num_stations;
 
+		/* In non-MLO/SLO case ah->num_radio is 1, and ar->mac_addr is
+		 * assigned to ieee80211_hw. In MLO with ah->num_radio > 1,
+		 * ab_dflt->mac_addr is assigned to ieee80211_hw.
+		 */
+		if (i == 0)
+			mac_addr = ar->mac_addr;
+		else
+			mac_addr = ab_dflt->mac_addr;
+
 		ar++;
 	}
 
@@ -15632,7 +15642,7 @@ static int ath12k_mac_hw_register(struct ath12k_hw *ah)
                         goto err_cleanup;
         }
 
-	SET_IEEE80211_PERM_ADDR(hw, ab_dflt->mac_addr);
+	SET_IEEE80211_PERM_ADDR(hw, mac_addr);
 	SET_IEEE80211_DEV(hw, ab_dflt->dev);
 
 	/* Iface modes are expected to be similar for partner chips */
-- 
2.34.1

